js/libtextsecure.js

```diff
@@ -15252,31 +15227,33 @@ window.axolotl.protocol = function() {
             return identityKeyCalculated(identityKeyPair);
     }
 
+	//TODO: Replace this stuff
     refreshPreKeys = function() {
         self.generateKeys().then(function(keys) {
             console.log("Pre Keys updated!");
-            return axolotl.api.updateKeys(keys);
+            return updateKeysCallback(keys);
         }).catch(function(e) {
             //TODO: Notify the user somehow???
             console.error(e);
         });
     }
 
-    window.setInterval(function() {
-        // Note that this will not ever run until generateKeys has been called at least once
-        if (axolotl.api.storage.get("lastPreKeyUpdate", Date.now()) < Date.now() - MESSAGE_LOST_THRESHOLD_MS)
-            refreshPreKeys();
-    }, 60 * 1000);
+	if (updateKeysCallback)
+		window.setInterval(function() {
+			// Note that this will not ever run until generateKeys has been called at least once
+			if (storage_interface.get("lastPreKeyUpdate", Date.now()) < Date.now() - MESSAGE_LOST_THRESHOLD_MS)
+				refreshPreKeys();
+		}, 60 * 1000);
```

Comment:
Yea, that interface is gonna get refactored a bit - the callback out to generate new keys should be triggered during message send/description, when we realize we need more, and then called in a promise asynchronously, not via a timer.

Matt

On February 16, 2015 12:22:12 AM PST, Lilia <notifications@github.com> wrote:
>>          }).catch(function(e) {
>>              //TODO: Notify the user somehow???
>>              console.error(e);
>>          });
>>      }
>>  
>> -    window.setInterval(function() {
>> -        // Note that this will not ever run until generateKeys has
>been called at least once
>> -        if (axolotl.api.storage.get("lastPreKeyUpdate", Date.now())
>< Date.now() - MESSAGE_LOST_THRESHOLD_MS)
>> -            refreshPreKeys();
>> -    }, 60 * 1000);
>> +	if (updateKeysCallback)
>> +		window.setInterval(function() {
>> +			// Note that this will not ever run until generateKeys has been
>called at least once
>> +			if (storage_interface.get("lastPreKeyUpdate", Date.now()) <
>Date.now() - MESSAGE_LOST_THRESHOLD_MS)
>> +				refreshPreKeys();
>> +		}, 60 * 1000);
>
>Just noticed that this setInterval will run in the background page as
>well as any frontend windows. We should only set this interval in the
>background page, if at all. It seems like this logic should live
>outside libaxolotl, which would eliminate the need for the
>updateKeysCallback...
>
>Also, checking once a minute and generating 100 keys a week sounds like
>overkill. Maybe we can rethink the numbers here.
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/WhisperSystems/TextSecure-Browser/pull/162/files#r24736280